### PR TITLE
#90 Add grouping by name to fetch_credentials()

### LINKED ISSUE #90: Receive credentials from the server
*Subtask of WoTTsecurity/api#86*
*Requires #89*

Schedule periodic GET request to `/api/.../creds` every 15 minutes. Write the result to a file.

Credentials are in the form of `[{"name": ..., "key": ..., "value": ...}, ...]`. Agent should write `{key: value, ...}` to a file named `{name}.json` (grouping credentials by `name`).

### PR DESCRIPTION
**Pull request by eugenosm**
Closes #90

### CODE REVIEW

File: agent/__init__.py [hunk 1]
```diff
@@ -404,14 +406,68 @@ def renew_expired_cert(csr, device_id, debug=False):
...
+def setup_endpoints(dev):
     if dev:
         global WOTT_ENDPOINT, MTLS_ENDPOINT, DASH_ENDPOINT
         endpoint = os.getenv('WOTT_ENDPOINT', 'http://localhost')
         DASH_ENDPOINT = endpoint + ':' + str(DASH_DEV_PORT)
         WOTT_ENDPOINT = endpoint + ':' + str(WOTT_DEV_PORT) + '/api'
         MTLS_ENDPOINT = endpoint + ':' + str(MTLS_DEV_PORT) + '/api'
 
+
+def fetch_creds(debug, dev):
```
Comment (vpetersson) [inline agent/__init__.py]: Don't abbreviate words in functions.

File: agent/__main__.py [hunk 2]
```diff
@@ -50,12 +50,19 @@ def main():
 
 PING_INTERVAL = 60 * 60
 PING_TIMEOUT = 10 * 60
+CREDS_INTERVAL = 15 * 60
```
Comment (vpetersson) [inline agent/__main__.py]: Same as above. Don't abbreviate.

File: agent/__init__.py [hunk 3]
```diff
@@ -44,6 +45,7 @@
 CA_CERT_PATH = os.path.join(CERT_PATH, 'ca.crt')
 COMBINED_PEM_PATH = os.path.join(CERT_PATH, 'combined.pem')
 INI_PATH = os.path.join(CONFIG_PATH, 'config.ini')
+CREDS_PATH = os.path.join(CONFIG_PATH, 'credentials')
```
Comment (vpetersson) [inline agent/__init__.py]: Same.

File: agent/__init__.py [hunk 4]
```diff
@@ -404,14 +406,68 @@ def renew_expired_cert(csr, device_id, debug=False):
...
+
+    by_names = {}
+    for cred in creds:
+        name = cred['name']
+        if name not in by_names:
+            by_names[name] = {}
+        by_names[name][cred['key']] = cred['value']
+
+    for name in by_names:
+        creds_fname = os.path.join(CREDS_PATH, "{}.json".format(name))
```
Comment (vpetersson) [inline agent/__init__.py]: This is not a filename, but a file path. Also, don't abbreviate this much as it makes the code harder to read. Readability trumps nifty tricks.

File: agent/__init__.py [hunk 5]
```diff
@@ -404,14 +406,68 @@ def renew_expired_cert(csr, device_id, debug=False):
...
+            print("[RECEIVED] Fetch creds: code {}".format(creds_req.status_code))
+            print("[RECEIVED] Fetch creds: {}".format(creds_req.content))
+        return
+    creds = creds_req.json()
+
+    print('Credentials retreived.')
+    if debug:
+        print('Creds: {}'.format(creds))
+
+    by_names = {}
```
Comment (vpetersson) [inline agent/__init__.py]: What does 'by_name" mean?

File: agent/__init__.py (hunk 1 above)
Comment (eugenosm) [reply agent/__init__.py]: noted

File: agent/__main__.py (hunk 2 above)
Comment (eugenosm) [reply agent/__main__.py]: +

File: agent/__init__.py (hunk 3 above)
Comment (eugenosm) [reply agent/__init__.py]: +

File: agent/__init__.py (hunk 4 above)
Comment (eugenosm) [reply agent/__init__.py]: noted

File: agent/__init__.py (hunk 5 above)
Comment (eugenosm) [reply agent/__init__.py]: the credentials was got in request are in array, so to make only one writes for each file resort it to dictionary by their 'name'...so may be not actually right var name

File: agent/__init__.py (hunk 3 above)
Comment (vpetersson) [reply agent/__init__.py]: This is still not resolved.

**Comment by a-martynovich** [comment]
@eugenosm I just tested your branch and found out that the folder residing in `CREDS_PATH` is not created if it does not exist. It should be created.

**Comment by eugenosm** [comment]
and what we'll do if there is file with the name of our credentials dir?
cry something and exit? or delete file and create dir?

**Comment by a-martynovich** [comment]
Tests look much better now. Only I would parse JSON [here](https://github.com/WoTTsecurity/agent/pull/107/files#diff-621c77fea2394a9a79831e44a3b482b2R518) and compare dicts instead of byte-strings, but that's splitting hairs.

**Comment by a-martynovich** [comment]
LGTM

**Pull request merged**